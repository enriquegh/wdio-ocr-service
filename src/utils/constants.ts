import { OcrServiceConfig } from '../types/types'

export const DEFAULT_OPTIONS: OcrServiceConfig = {
  ocrImagesPath: './.tmp',
}
export const SERVICE_NAME = require('../../package.json').name
