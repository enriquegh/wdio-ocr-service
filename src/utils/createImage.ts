// @ts-ignore
import { encodePNGToStream, decodePNGFromStream, make } from 'pureimage'
import { createReadStream, createWriteStream } from 'fs'
import { Rectangles } from '../types/types'

interface CreateImageData {
  filePath: string;
  height: number;
  lines: any;
  width: number;
}

export default async function createImage(data: CreateImageData) {
  const { filePath, height, lines, width } = data

  const image = await decodePNGFromStream(createReadStream(filePath))
  const canvasImage = make(width, height)
  const context = canvasImage.getContext('2d')
  context.drawImage(
    image,
    // Start at x/y pixels from the left and the top of the image (crop)
    0, 0,
    // 'Get' a (w * h) area from the source image (crop)
    width, height,
    // Place the result at 0, 0 in the canvas,
    0, 0,
    // With as width / height: 100 * 100 (scale)
    width, height
  )

  // Highlight all found texts
  lines.forEach(({ bbox }: { bbox: Rectangles }) => {
    const { right, bottom, left, top } = bbox
    context.beginPath()
    context.fillStyle = 'rgba(57, 170, 86, 0.5)'
    context.fillRect(left, top, right - left, bottom - top)
    context.lineWidth = 2
    context.strokeStyle = '#39aa56'
    context.rect(left, top, right - left, bottom - top)
    context.stroke()
  })

  await encodePNGToStream(canvasImage, createWriteStream(filePath))
}
