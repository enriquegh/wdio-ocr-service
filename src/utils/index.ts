import {
  ClickPoint,
  DetermineClickPointOptions,
  Rectangles,
  ScreenSize,
} from '../typings/types'

/**
 * Get the size of a screenshot in pixels
 */
export function getScreenshotSize(screenshot: string): ScreenSize {
  return {
    height: Buffer.from(screenshot, 'base64').readUInt32BE(20),
    width: Buffer.from(screenshot, 'base64').readUInt32BE(16),
  }
}

export function getDprPositions(values: Rectangles, dpr: number): Rectangles {
  Object.keys({ ...values }).map((value: string) => {
    // @ts-ignore
    values[value] /= dpr
  })

  return values
}

/**
 * Determine the click point
 */
export function determineClickPoint(options: DetermineClickPointOptions): ClickPoint {
  const { rectangles: { left, right, top, bottom } } = options
  const x = left + (right - left) / 2
  const y = top + (bottom - top) / 2

  return { x, y }
}
