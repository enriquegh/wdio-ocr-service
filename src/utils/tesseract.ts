import { exec, ExecException } from 'child_process'
import { createWorker, OEM, PSM } from 'tesseract.js'
// @ts-ignore
import { recognize } from 'node-tesseract-ocr'
import { parseString } from 'xml2js'
import { GetOcrData, Line, Rectangles, Words } from '../types/types'

export async function isTesseractAvailable(
  tesseractName: string = ''
): Promise<boolean> {
  const binary = tesseractName || 'tesseract'
  const command = [binary, '--version'].join(' ')

  return new Promise((resolve) => {
    exec(command, (error: ExecException | null) => {
      if (error) {
        return resolve(false)
      }

      return resolve(true)
    })
  })
}

interface GetOcrDataOptions {
  filePath: string;
}

export async function getNodeOcrData(options: GetOcrDataOptions): Promise<GetOcrData|Error> {
  try {
    const { filePath } = options
    const jsonSingleWords: Words[] = []
    const jsonWordStrings: Line[] = []
    let composedBlocks: any = []

    const worker = createWorker()
    await worker.load()
    await worker.loadLanguage('eng')
    await worker.initialize('eng')
    await worker.setParameters({
      tessedit_ocr_engine_mode: OEM.TESSERACT_LSTM_COMBINED,
      tessedit_pageseg_mode: PSM.AUTO,
      tessjs_create_tsv: '0',
      tessjs_create_box: '0',
      tessjs_create_unlv: '0',
      tessjs_create_osd: '0',
    })
    const { data: { text, hocr } } = await worker.recognize(filePath)

    // @ts-ignore
    parseString(hocr, (error: Error, data: any) => {
      if (error) {
        throw Error(`An error happened when parsing the getNodeOcrData, see: ${error}`)
      }

      composedBlocks = data.div.div
    })

    if (!composedBlocks || composedBlocks.length === 0){
      throw Error('No text was found for the OCR, please verify the stored image.')
    }

    // This is for single words
    // @ts-ignore
    composedBlocks.forEach(({ p: TextBlock }) => {
      // @ts-ignore
      TextBlock.forEach(({ span: TextLine }) => {
        // @ts-ignore
        TextLine.forEach(({ span: String }) => {
          // @ts-ignore
          String.forEach(({ _: text, $: { title } }) => {
            if (!text) {
              return
            }

            const attributes = `; ${title}`.split('; ')
            const { bbox, wc } = parseAttributeString(attributes)

            jsonSingleWords.push({
              text,
              bbox,
              wc,
            })
          })
        })
      })
    })

    // This is for single lines
    // @ts-ignore
    composedBlocks.forEach(({ p: TextBlock }) => {
      // @ts-ignore
      TextBlock.forEach(({ span: TextLine }) => {
        // @ts-ignore
        TextLine.forEach(({ $: { title }, span: String }) => {
          const attributes = `; ${title}`.split('; ')
          const { bbox } = parseAttributeString(attributes)
          const line = {
            text: '',
            bbox,
          }

          // @ts-ignore
          String.map(({ _: text }) => {
            line.text = `${line.text} ${text || ''}`.trim()
          })

          if (line.text === '') {
            return
          }

          jsonWordStrings.push(line)
        })
      })
    })

    await worker.terminate()

    return {
      lines: jsonWordStrings,
      words: jsonSingleWords,
      text,
    }
  } catch (error) {
    throw Error(`An error happened when parsing the getNodeOcrData, see: ${error}`)
  }
}

export async function getSystemOcrData(options: GetOcrDataOptions): Promise<GetOcrData|Error> {
  try {
    const { filePath } = options
    const jsonSingleWords: Words[] = []
    const jsonWordStrings: Line[] = []
    let composedBlocks: any = []
    let text: string = ''
    const result = await recognize(filePath, {
      lang: 'eng',
      oem: 1,
      // https://github.com/tesseract-ocr/tesseract/blob/master/doc/tesseract.1.asc
      psm: 3,
      presets: ['txt', 'alto'],
    })

    parseString(result, (error: Error, data) => {
      if (error) {
        throw Error(`An error happened when parsing the getSystemOcrData, see: ${error}`)
      }

      text = data.alto.Layout[0]._
      composedBlocks = data.alto.Layout[0].Page[0].PrintSpace[0].ComposedBlock
    })

    if (!composedBlocks || composedBlocks.length === 0){
      throw Error('No text was found for the OCR, please verify the stored image.')
    }

    // This is for single words
    // @ts-ignore
    composedBlocks.forEach(({ TextBlock }) => {
      // @ts-ignore
      TextBlock.forEach(({ TextLine }) => {
        // @ts-ignore
        TextLine.forEach(({ String }) => {
          // @ts-ignore
          String.forEach(({ $: { CONTENT, HPOS, VPOS, WIDTH, HEIGHT, WC } }) => {
            jsonSingleWords.push({
              text: CONTENT || '',
              bbox: {
                left: Number(HPOS),
                top: Number(VPOS),
                right: Number(HPOS) + Number(WIDTH),
                bottom: Number(VPOS) + Number(HEIGHT),
              },
              wc: Number(WC),
            })
          }
          )
        })
      })
    })

    // This is for single lines
    // @ts-ignore
    composedBlocks.forEach(({ TextBlock }) => {
      // @ts-ignore
      TextBlock.forEach(({ TextLine }) => {
        // @ts-ignore
        TextLine.forEach(({ $: { HPOS, VPOS, WIDTH, HEIGHT }, String }) => {
          const line = {
            text: '',
            bbox: {
              left: Number(HPOS),
              top: Number(VPOS),
              right: Number(HPOS) + Number(WIDTH),
              bottom: Number(VPOS) + Number(HEIGHT),
            },
          }
          // @ts-ignore
          String.forEach(({ $: { CONTENT } }) => {
            line.text = `${line.text} ${CONTENT || ''}`.trim()
          })

          if (line.text === '') {
            return
          }

          jsonWordStrings.push(line)
        })
      })
    })

    return {
      lines: jsonWordStrings,
      words: jsonSingleWords,
      text,
    }
  } catch (error) {
    throw Error(`An error happened when parsing the getSystemOcrData, see: ${error}`)
  }
}

export function parseAttributeString(
  attributes: string[]
): { bbox: Rectangles; wc: number } {
  let bbox = {
    left: 0,
    top: 0,
    right: 0,
    bottom: 0,
  }
  let wc = 0

  attributes.forEach((attribute: string) => {
    if (attribute.includes('bbox')) {
      const bboxValues = attribute.replace('bbox ', '').split(' ')
      bbox = {
        left: Number(bboxValues[0]),
        top: Number(bboxValues[1]),
        right: Number(bboxValues[2]),
        bottom: Number(bboxValues[3]),
      }
    } else if (attribute.includes('x_wconf')) {
      const score = attribute.replace('x_wconf ', '')
      wc = Number(score) / 100
    }
  })

  return {
    ...{ bbox },
    wc,
  }
}
