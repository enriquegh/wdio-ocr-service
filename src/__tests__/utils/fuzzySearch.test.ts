import Fuse from 'fuse.js'
import {fuzzyFind} from '../../utils/fuzzySearch'

jest.mock('fuse.js')

describe('utils - fuzzySearch', () => {
  it('should find data when only 1 result is given back with a 100% match', () => {
    (Fuse as unknown as jest.Mock).mockImplementation(() => ({
      search: jest.fn().mockReturnValue([
        {
          item: {
            text: 'Username',
            originalPosition: {left: 1, top: 2, right: 3, bottom: 4},
            dprPosition: {left: 1, top: 2, right: 3, bottom: 4},
          },
          refIndex: 0,
          score: 2.220446049250313e-16
        },
      ])
    }))

    const options = {
      textArray: [
        {
          text: 'Username',
          originalPosition: {left: 1, top: 2, right: 3, bottom: 4},
          dprPosition: {left: 1, top: 2, right: 3, bottom: 4},
        },
      ],
      pattern: 'Username',
    }

    expect(fuzzyFind(options)).toMatchSnapshot()
    expect(Fuse).toHaveBeenCalledWithSnapshot()
  })

  it('should find data when multiple results are given back', () => {
    (Fuse as unknown as jest.Mock).mockImplementation(() => ({
      search: jest.fn().mockReturnValue([
        {
          item: {
            text: 'Username',
            originalPosition: {left: 1, top: 2, right: 3, bottom: 4},
            dprPosition: {left: 1, top: 2, right: 3, bottom: 4},
          },
          refIndex: 0,
          score: 2.220446049250313e-16
        },
        {
          item: {
            text: 'The currently accepted usernames for',
            originalPosition: {left: 1, top: 2, right: 3, bottom: 4},
            dprPosition: {left: 1, top: 2, right: 3, bottom: 4},
          },
          refIndex: 5,
          score: 0.5184326474378735
        },
      ])
    }))

    const options = {
      textArray: [
        {
          text: 'Username',
          originalPosition: {left: 1, top: 2, right: 3, bottom: 4},
          dprPosition: {left: 1, top: 2, right: 3, bottom: 4},
        },
        {
          text: 'Password',
          originalPosition: {left: 1, top: 2, right: 3, bottom: 4},
          dprPosition: {left: 1, top: 2, right: 3, bottom: 4},
        },
        {
          text: 'LOGIN',
          originalPosition: {left: 1, top: 2, right: 3, bottom: 4},
          dprPosition: {left: 1, top: 2, right: 3, bottom: 4},
        },
      ],
      pattern: 'Username',
    }

    expect(fuzzyFind(options)).toMatchSnapshot()
    expect(Fuse).toHaveBeenCalledWithSnapshot()
  })
})
