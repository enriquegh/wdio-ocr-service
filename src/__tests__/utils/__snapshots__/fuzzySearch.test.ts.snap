// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils - fuzzySearch should find data when multiple results are given back 1`] = `
Array [
  Object {
    "item": Object {
      "dprPosition": Object {
        "bottom": 4,
        "left": 1,
        "right": 3,
        "top": 2,
      },
      "originalPosition": Object {
        "bottom": 4,
        "left": 1,
        "right": 3,
        "top": 2,
      },
      "text": "Username",
    },
    "refIndex": 0,
    "score": 0,
  },
  Object {
    "item": Object {
      "dprPosition": Object {
        "bottom": 4,
        "left": 1,
        "right": 3,
        "top": 2,
      },
      "originalPosition": Object {
        "bottom": 4,
        "left": 1,
        "right": 3,
        "top": 2,
      },
      "text": "The currently accepted usernames for",
    },
    "refIndex": 5,
    "score": 0.5184326474378735,
  },
]
`;

exports[`utils - fuzzySearch should find data when multiple results are given back 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [
        Object {
          "dprPosition": Object {
            "bottom": 4,
            "left": 1,
            "right": 3,
            "top": 2,
          },
          "originalPosition": Object {
            "bottom": 4,
            "left": 1,
            "right": 3,
            "top": 2,
          },
          "text": "Username",
        },
      ],
      Object {
        "ignoreFieldNorm": false,
        "ignoreLocation": false,
        "includeMatches": false,
        "includeScore": true,
        "isCaseSensitive": false,
        "keys": Array [
          "text",
        ],
        "shouldSort": true,
        "useExtendedSearch": false,
      },
    ],
    Array [
      Array [
        Object {
          "dprPosition": Object {
            "bottom": 4,
            "left": 1,
            "right": 3,
            "top": 2,
          },
          "originalPosition": Object {
            "bottom": 4,
            "left": 1,
            "right": 3,
            "top": 2,
          },
          "text": "Username",
        },
        Object {
          "dprPosition": Object {
            "bottom": 4,
            "left": 1,
            "right": 3,
            "top": 2,
          },
          "originalPosition": Object {
            "bottom": 4,
            "left": 1,
            "right": 3,
            "top": 2,
          },
          "text": "Password",
        },
        Object {
          "dprPosition": Object {
            "bottom": 4,
            "left": 1,
            "right": 3,
            "top": 2,
          },
          "originalPosition": Object {
            "bottom": 4,
            "left": 1,
            "right": 3,
            "top": 2,
          },
          "text": "LOGIN",
        },
      ],
      Object {
        "ignoreFieldNorm": false,
        "ignoreLocation": false,
        "includeMatches": false,
        "includeScore": true,
        "isCaseSensitive": false,
        "keys": Array [
          "text",
        ],
        "shouldSort": true,
        "useExtendedSearch": false,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "search": [MockFunction] {
          "calls": Array [
            Array [
              "Username",
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": Array [
                Object {
                  "item": Object {
                    "dprPosition": Object {
                      "bottom": 4,
                      "left": 1,
                      "right": 3,
                      "top": 2,
                    },
                    "originalPosition": Object {
                      "bottom": 4,
                      "left": 1,
                      "right": 3,
                      "top": 2,
                    },
                    "text": "Username",
                  },
                  "refIndex": 0,
                  "score": 0,
                },
              ],
            },
          ],
        },
      },
    },
    Object {
      "type": "return",
      "value": Object {
        "search": [MockFunction] {
          "calls": Array [
            Array [
              "Username",
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": Array [
                Object {
                  "item": Object {
                    "dprPosition": Object {
                      "bottom": 4,
                      "left": 1,
                      "right": 3,
                      "top": 2,
                    },
                    "originalPosition": Object {
                      "bottom": 4,
                      "left": 1,
                      "right": 3,
                      "top": 2,
                    },
                    "text": "Username",
                  },
                  "refIndex": 0,
                  "score": 0,
                },
                Object {
                  "item": Object {
                    "dprPosition": Object {
                      "bottom": 4,
                      "left": 1,
                      "right": 3,
                      "top": 2,
                    },
                    "originalPosition": Object {
                      "bottom": 4,
                      "left": 1,
                      "right": 3,
                      "top": 2,
                    },
                    "text": "The currently accepted usernames for",
                  },
                  "refIndex": 5,
                  "score": 0.5184326474378735,
                },
              ],
            },
          ],
        },
      },
    },
  ],
}
`;

exports[`utils - fuzzySearch should find data when only 1 result is given back with a 100% match 1`] = `
Array [
  Object {
    "item": Object {
      "dprPosition": Object {
        "bottom": 4,
        "left": 1,
        "right": 3,
        "top": 2,
      },
      "originalPosition": Object {
        "bottom": 4,
        "left": 1,
        "right": 3,
        "top": 2,
      },
      "text": "Username",
    },
    "refIndex": 0,
    "score": 0,
  },
]
`;

exports[`utils - fuzzySearch should find data when only 1 result is given back with a 100% match 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [
        Object {
          "dprPosition": Object {
            "bottom": 4,
            "left": 1,
            "right": 3,
            "top": 2,
          },
          "originalPosition": Object {
            "bottom": 4,
            "left": 1,
            "right": 3,
            "top": 2,
          },
          "text": "Username",
        },
      ],
      Object {
        "ignoreFieldNorm": false,
        "ignoreLocation": false,
        "includeMatches": false,
        "includeScore": true,
        "isCaseSensitive": false,
        "keys": Array [
          "text",
        ],
        "shouldSort": true,
        "useExtendedSearch": false,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "search": [MockFunction] {
          "calls": Array [
            Array [
              "Username",
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": Array [
                Object {
                  "item": Object {
                    "dprPosition": Object {
                      "bottom": 4,
                      "left": 1,
                      "right": 3,
                      "top": 2,
                    },
                    "originalPosition": Object {
                      "bottom": 4,
                      "left": 1,
                      "right": 3,
                      "top": 2,
                    },
                    "text": "Username",
                  },
                  "refIndex": 0,
                  "score": 0,
                },
              ],
            },
          ],
        },
      },
    },
  ],
}
`;
